crypto/core_namemap.c

```diff
@@ -216,62 +179,97 @@ int ossl_namemap_name2num_n(const OSSL_NAMEMAP *namemap,
+        return NULL;
+
+    names = sk_NAMES_value(namemap->numnames, number - 1);
+    if (names != NULL)
+        ret = sk_STRING_value(names, idx);
+
+    CRYPTO_THREAD_unlock(namemap->lock);
+
+    return ret;
 }
 
-const char *ossl_namemap_num2name(const OSSL_NAMEMAP *namemap, int number,
-                                  size_t idx)
+/* This function is not thread safe, the namemap must be locked */
+static int numname_insert(OSSL_NAMEMAP *namemap, int number,
+                          const char *name)
 {
-    struct num2name_data_st data;
+    NAMES *names;
+    char *tmpname;
+
+    if (number > 0) {
+        names = sk_NAMES_value(namemap->numnames, number - 1);
+        if (!ossl_assert(names != NULL)) {
+            /* cannot happen */
+            return 0;
+        }
+    } else {
+        /* a completely new entry */
+        names = sk_STRING_new_null();
+        if (names == NULL)
+            return 0;
+    }
 
-    data.idx = idx;
-    data.name = NULL;
-    if (!ossl_namemap_doall_names(namemap, number, do_num2name, &data))
-        return NULL;
-    return data.name;
+    if ((tmpname = OPENSSL_strdup(name)) == NULL)
+        goto err;
+
+    if (!sk_STRING_push(names, tmpname))
+        goto err;
+
+    if (number <= 0) {
+        if (!sk_NAMES_push(namemap->numnames, names))
+            goto err;
+        number = sk_NAMES_num(namemap->numnames);
+    }
+    return number;
+
+ err:
+    if (number <= 0)
+        sk_STRING_free(names);
+    OPENSSL_free(tmpname);
+    return 0;
 }
 
 /* This function is not thread safe, the namemap must be locked */
 static int namemap_add_name(OSSL_NAMEMAP *namemap, int number,
                             const char *name)
 {
-    NAMENUM_ENTRY *namenum = NULL;
-    int tmp_number;
+    int ret;
+    HT_VALUE val = { 0 };
+    NAMENUM_KEY key;
 
     /* If it already exists, we don't add it */
-    if ((tmp_number = namemap_name2num(namemap, name)) != 0)
-        return tmp_number;
+    if ((ret = ossl_namemap_name2num(namemap, name)) != 0)
+        return ret;
 
-    if ((namenum = OPENSSL_zalloc(sizeof(*namenum))) == NULL)
+    if ((number = numname_insert(namemap, number, name)) == 0)
         return 0;
 
-    if ((namenum->name = OPENSSL_strdup(name)) == NULL)
-        goto err;
-
-    /* The tsan_counter use here is safe since we're under lock */
-    namenum->number =
-        number != 0 ? number : 1 + tsan_counter(&namemap->max_number);
-    (void)lh_NAMENUM_ENTRY_insert(namemap->namenum, namenum);
-
-    if (lh_NAMENUM_ENTRY_error(namemap->namenum))
-        goto err;
-    return namenum->number;
+    /* Using tsan_store alone here is safe since we're under lock */
+    tsan_store(&namemap->max_number, number);
 
- err:
-    namenum_free(namenum);
-    return 0;
+    HT_INIT_KEY(&key);
+    HT_SET_KEY_STRING_CASE(&key, name, name);
+    val.value = (void *)(intptr_t)number;
+    ret = ossl_ht_insert(namemap->namenum_ht, TO_HT_KEY(&key), &val, NULL);
```

Comment:
The NULL parameter here indicates this is a non-replacement operation meaning that, if this fails, there is already a matching value in the hash table.  If we move the push to the name stack below this, then we don't need to do the pre-check of the mappings existence with the call to ossl_namemap_name2num above.